### PR TITLE
feat(website): submit filters with one button click

### LINKED ISSUE #35: Submit button for search requires 2 actions instead of 1 to submit - this is confusing
When I am searching for sequencing I keep forgetting that I need to enter and then still click submit to search - normally this is one action. It is especially annoying as if I forget to enter before pressing submit, then the page reloads and resets my filters (e.g. in the percentage of mutations) and I have to start from the beginning in my search for variants of interest:
image

### PR DESCRIPTION
**Pull request by JonasKellerer**
Resolves #35

resolves #35

### Summary

The construction of the filters for baseline and variants is now performed client-side. This way, we no longer need different buttons to filter for each of them. 

In this PR, I also use the new way to provide date ranges to the range selector component.

### Screenshot

Single variant page
grafik

Sequencing efforts page
grafik

Compare side by side page
grafik

### CODE REVIEW

File: website/src/components/pageStateSelectors/ApplyFilterButton.tsx
Comment (fengelniederhammer) [inline website/src/components/pageStateSelectors/ApplyFilterButton.tsx:1]: Also here component name and filename don't match

File: website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        };
+        window.location.href = view.toUrl(newPageState);
+    };
+
+    return (
+        <div className='flex flex-col gap-6 bg-gray-50 p-2'>
+            <BaselineSelector
+                onLocationChange={(location) => setLocation(location)}
+                locationFilterConfig={locationFilterConfig}
+                onDateRangeChange={(dateRange) => setDateRange(dateRange)}
```
Comment (fengelniederhammer) [inline website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx:75]:
```suggestion
                onLocationChange={setLocation}
                locationFilterConfig={locationFilterConfig}
                onDateRangeChange={setDateRange}
```

File: website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx
Comment (fengelniederhammer) [inline website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx:1]: I hit a state where the date range is somehow invalid:
grafik

I can't get back to a normal state (since I can't use the date selector anymore) without navigating to the page again or manipulation the URL (reloading doesn't help).

-> I think a "reset" button on this component would make sense.

File: website/src/pages/covid/single-variant.astro [hunk 2]
```diff
@@ -39,38 +30,37 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
+                        initialValue: pageState.variantFilter.lineage,
+                        lapisField: view.lineageField,
+                        placeholderText: 'Pango Lineage',
+                    }}
+                    pageState={pageState}
+                    organismViewKey={organismViewKey}
+                    organismsConfig={getDashboardsConfig().dashboards.organisms}
+                    client:only='react'
+                />
+                <!--<div class='divider'></div>-->
```
Comment (fengelniederhammer) [inline website/src/pages/covid/single-variant.astro:57]:
```suggestion
```

File: website/src/pages/covid/single-variant.astro [hunk 3]
```diff
@@ -39,38 +30,37 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
+                <div class='p-2'>
+                    <h2 class='my-2'>Or select a known variant:</h2>
                     <CollectionsList
                         initialCollectionId={pageState.collectionId}
                         organismsConfig={getDashboardsConfig().dashboards.organisms}
-                        client:idle
+                        client:only='react'
```
Comment (fengelniederhammer) [inline website/src/pages/covid/single-variant.astro:63]: Why did you change this?

(Can't we actually use `client:load` here?)

File: website/src/pages/covid/single-variant.astro [hunk 4]
```diff
@@ -39,38 +30,37 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
+                        initialDateRange: pageState.baselineFilter.dateRange,
+                        dateRangeOptions: view.dateRangeOptions,
+                        earliestDate: view.earliestDate,
+                        dateColumn: view.mainDateField,
+                    }}
+                    mutationFilterConfig={{ initialMutations }}
+                    lineageFilterConfig={{
+                        initialValue: pageState.variantFilter.lineage,
+                        lapisField: view.lineageField,
+                        placeholderText: 'Pango Lineage',
```
Comment (fengelniederhammer) [inline website/src/pages/covid/single-variant.astro:50]: Isn't this actually the Nextclade pango lineage?

(We could also compute the placholder text from `view.lineageField` by using sentence case or similar)

File: website/src/pages/flu/h5n1/single-variant.astro [hunk 5]
```diff
@@ -33,80 +29,81 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
-                        organismViewKey={organismViewKey}
-                    />
-                </div>
-                <div class='border-t-2 bg-green-50 px-2 py-4'>
-                    <div class='mb-2 font-semibold'>Search variant:</div>
-                    <VariantFilter fields={[]} initialMutations={initialMutations} organismViewKey={organismViewKey} />
-                </div>
-            </div>
-            <div>
-                <ComponentsGrid>
```
Comment (fengelniederhammer) [inline website/src/pages/flu/h5n1/single-variant.astro:54]: Does removing this div break the grid layout?

This is how it looks now: 
grafik

This is how it should look like:
grafik

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]

File: website/src/views/covid.ts [hunk 6]
```diff
@@ -25,9 +23,13 @@ const earliestDate = '2020-01-06';
 
 class CovidConstants {
     public readonly organism = Organisms.covid;
-    public readonly defaultDateRange: DateRange = 'last6Months';
+    public readonly defaultDateRange: DateRangeOption = {
```
Comment (fengelniederhammer) [inline website/src/views/covid.ts:26]: We could also use `DateRangeOption` from `@genspectrum/dashboard-components` (because ultimately, that's the type that counts)

File: website/src/views/helpers.ts
Comment (fengelniederhammer) [inline website/src/views/helpers.ts:1]: We don't need the `SpecialDateRange` anymore, do we?

File: website/src/views/helpers.ts [hunk 7]
```diff
@@ -60,15 +60,9 @@ export const dateRangeToCustomDateRange = (dateRange: DateRange, allTimesStartDa
     };
 };
 
-export const chooseGranularityBasedOnDateRange = (
-    dateRange: DateRange,
-    allTimesStartDate: Date,
-): 'day' | 'week' | 'month' | 'year' => {
-    if (!isCustomDateRange(dateRange)) {
-        dateRange = dateRangeToCustomDateRange(dateRange, allTimesStartDate);
```
Comment (fengelniederhammer) [inline website/src/views/helpers.ts:68]: `dateRangeToCustomDateRange` seems unused now

File: website/src/views/helpers.ts [hunk 8]
```diff
@@ -102,14 +96,14 @@ export const setSearchFromStringArray = (search: URLSearchParams, name: string,
...
     search: URLSearchParams,
     name: string,
-    dateRange: DateRange | undefined | null,
+    dateRange: DateRangeOption | undefined | null,
 ) => {
     if (dateRange !== null && dateRange !== undefined) {
         let serializedValue: string;
-        if (isCustomDateRange(dateRange)) {
-            serializedValue = `${dateRange.from}--${dateRange.to}`;
+        if (dateRange.label === 'Custom') {
```
Comment (fengelniederhammer) [inline website/src/views/helpers.ts]: Let's use a constant for 'Custom'

File: website/src/views/helpers.ts [hunk 9]
```diff
@@ -137,24 +131,29 @@ export const getIntegerFromSearch = (search: URLSearchParams, name: string): num
...
+        return customDateRange;
+    }
+
     if (value.includes('--')) {
-        // TODO Properly validate the value
         const split = value.split('--');
         return {
-            from: split[0],
-            to: split[1],
+            label: 'Custom',
```
Comment (fengelniederhammer) [inline website/src/views/helpers.ts]: Let's use a constant for 'Custom'

File: website/src/routes/View.ts
Comment (fengelniederhammer) [inline website/src/routes/View.ts:1]: We don't need this empty file ;)

File: website/src/pages/west-nile/single-variant.astro [hunk 10]
```diff
@@ -39,98 +30,92 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
+                        lapisField: view.lineageField,
+                        placeholderText: 'Lineage',
+                    }}
+                    pageState={pageState}
+                    organismViewKey={organismViewKey}
+                    organismsConfig={getDashboardsConfig().dashboards.organisms}
+                    client:only='react'
+                />
             </div>
-            <div>
```
Comment (fengelniederhammer) [inline website/src/pages/west-nile/single-variant.astro:70]: Same here: I also suspect that removing this div breakes the grid layout

File: website/src/components/pageStateSelectors/VariantSelector.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                    <GsTextInput
+                        lapisField={cladeFilterConfig.lapisField}
+                        placeholderText={cladeFilterConfig.placeholderText}
+                        onInputChange={(clade) => onCladeChange(clade[cladeFilterConfig.lapisField])}
+                    />
+                )}
+                <GsMutationFilter
+                    initialValue={mutationFilter}
+                    onMutationChange={onMutationChange}
+                    onBlur={onMutationChange}
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply website/src/components/pageStateSelectors/VariantSelector.tsx]: True. We don't need this here.

File: website/src/components/pageStateSelectors/VariantSelector.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                        onLineageChange={(lineage) => onLineageChange(lineage[lineageFilterConfig.lapisField])}
+                        initialValue={lineageFilterConfig.initialValue}
+                    />
+                )}
+                {cladeFilterConfig && (
+                    <GsTextInput
+                        lapisField={cladeFilterConfig.lapisField}
+                        placeholderText={cladeFilterConfig.placeholderText}
+                        onInputChange={(clade) => onCladeChange(clade[cladeFilterConfig.lapisField])}
+                    />
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply website/src/components/pageStateSelectors/VariantSelector.tsx]: True. It was also missing inside the `GsTextInput` component.

File: website/src/pages/flu/h5n1/single-variant.astro (hunk 5 above)
Comment (JonasKellerer) [reply website/src/pages/flu/h5n1/single-variant.astro]: I changed it, so the mutations over time components are again each on a single row.

File: website/src/views/covid.ts (hunk 6 above)
Comment (JonasKellerer) [reply website/src/views/covid.ts:26]: I now used it, and removed in the same step also the 'todayString'

File: website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx
Comment (JonasKellerer) [reply website/src/components/pageStateSelectors/SingleVariantPageStateSelector.tsx:1]: One part was fixed, by using the correct dates in the filter #276. The reset button will be covered by #285.

File: website/src/pages/covid/single-variant.astro (hunk 4 above)
Comment (JonasKellerer) [reply website/src/pages/covid/single-variant.astro]: There is now #292, which looks at the specification of the filters in the config and #293, which looks at deriving the placeholder text from the lapis field name.

File: website/src/pages/mpox/single-variant.astro [hunk 13]
```diff
@@ -39,42 +29,34 @@ const initialMutations = getLapisMutations(pageState.variantFilter);
...
+                cladeFilterConfig={{
+                    initialValue: pageState.variantFilter.clade,
+                    lapisField: view.cladeField,
+                    placeholderText: 'Clade',
+                }}
+                lineageFilterConfig={{
+                    initialValue: pageState.variantFilter.lineage,
+                    lapisField: view.lineageField,
+                    placeholderText: 'Lineage',
+                }}
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply website/src/pages/mpox/single-variant.astro]: I changed it.

**Review by fengelniederhammer — APPROVED** [review]
LGTM :+1:

File: website/src/components/rsv/RsvSequencingEffortsPage.astro [hunk 14]
```diff
@@ -36,71 +37,77 @@ const { label: subdivisionLabel, field: subdivisionField } = getLocationSubdivis
...
-                    fields={view.locationFields}
-                    initialLocation={pageState.baselineFilter.location}
-                    initialDateRange={pageState.baselineFilter.dateRange}
-                    earliestDate={view.earliestDate}
-                    customDateRangeOptions={view.customDateRangeOptions}
-                    organismViewKey={organismViewKey}
-                />
-            </div>
-        </div>
+        <div class='grid-cols-[300px_1fr] gap-x-4 sm:grid'>
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply website/src/components/rsv/RsvSequencingEffortsPage.astro]: I create a layout for all those pages.
